consensus/safety-rules/src/safety_rules.rs

```diff
@@ -98,6 +99,37 @@ impl<T: Payload> SafetyRules<T> {
         }
     }
 
+    /// This verifies the public key and private key of a validator against validator set
+    /// during epoch changes. We might want to add more detailed error code for identifying
+    /// the failure of key verification
+    fn verify_key(&mut self, epoch_state: &EpochState) -> Result<(), Error> {
+        // the validator set
+        let validator_verifier = &epoch_state.verifier;
+        let validator_signer = self
+            .validator_signer
+            .as_ref()
+            .ok_or(Error::NotInitialized)?;
+
+        // validator account not in the validator set
+        let public_key = validator_signer.public_key();
+        let author = validator_signer.author();
+        let validator_key = validator_verifier
+            .get_public_key(&author)
+            .ok_or(Error::InvalidAuthor)?;
+
+        // validator public key does not match
+        if public_key != validator_key {
+            let consensus_key = self
+                .persistent_storage
+                .consensus_key()
+                .expect("Unable to retrieve consensus private key");
+            self.validator_signer = Some(ValidatorSigner::new(author, consensus_key));
+            // Maybe we want to signal the safety rule with Error::InvalidPublicKey
+            return Err(Error::InvalidPublicKey);
```

Comment:
this isn't the right API, we need to request the consensus key given our current public key, look at https://github.com/libra/libra/blob/master/secure/storage/src/crypto_storage.rs#L45

Reply:
As David says, we should call "export_private_key_for_version(..)" using the consensus key name and the public key currently registered in the epoch change. This is because this code may still fetch a consensus private key that does not match the one in the epoch change.

Reply:
Thanks for the pointer to the right API! I'm also scratching my head in writing the tests for this part -- to test, safety rules must see epoch changes to update() while current tests under safety_rules seem "static": constructing proposal and checking a few properties, which do not update any epoch. Should I go look for other existing tests that triggers epoch changes for safety rules to update?